### PR TITLE
[refactor][AssetCondition] Begin moving AutoMaterializeRule implementations into AssetCondition

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

As it says in the title -- this begins the process of shuttling over the implementations of these rules into native AssetCondition objects. There's not particular urgency to move over all of these, but this sets up the pattern going forward.

Other changes may end up leading to the composition of multiple AssetConditions to replace a single rule.

### CODE REVIEW

**Comment by OwenKephart** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19750?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `master`
  * **PR #19553** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19553?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #19746** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19746?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #19747** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19747?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #19749** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19749?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #19750** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19750?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/19750?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 1]
```diff
@@ -1058,39 +1055,14 @@ def decision_type(self) -> AutoMaterializeDecisionType:
...
-        self, context: AssetConditionEvaluationContext
-    ) -> "AssetConditionResult":
-        from .asset_condition.asset_condition import AssetConditionResult
-
-        asset_partitions_by_evaluation_data = defaultdict(set)
-
-        subset_to_evaluate = (
-            context.candidates_not_evaluated_on_previous_tick_subset
-            | context.candidate_parent_has_or_will_update_subset
+    def to_asset_condition(self) -> AssetCondition:
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:1060]: probably missed this somewhere else, but how are we forking between to_asset_condition and evaluate_for_asset ?

**Review by alangenfeld — APPROVED** [review]
trusting tests that things were moved around without error

**Comment by OwenKephart** [comment]
@alangenfeld rethought the order that these changes should happen in, so closing this for now, but to answer your question we've been evaluating everything as AssetConditions natively for the past few releases (so AutoMaterializePolicy.to_asset_condition() gets called before any evaluation happens).

At the moment, most rules just get turned into RuleCondition(<some rule>), which then immediately invokes `evaluate_for_asset`, so it's still basically just executing a rule under the hood, but there's no requirement that there be a RuleCondition at the bottom of a tree

**Pull request closed without merging**